python/cudf/cudf/_lib/interop.pyx

```diff
@@ -102,7 +102,7 @@ cdef vector[column_metadata] gather_metadata(dict cols_dtypes) except *:
     if cols_dtypes is not None:
         for idx, (col_name, col_dtype) in enumerate(cols_dtypes.items()):
```

Comment:
`dict`'s starting python 3.6 are ordered, so this shouldn't be an issue as our minimum version requirements for python is `3.7`: https://stackoverflow.com/a/39980744

Reply:
We're not really using the dict-like aspect of the container though, but rather just its behaviour as an iterable of pairs. It's not important for this fix, but let me propose something in a followup.

Reply:
(Something like this: https://github.com/wence-/cudf/commit/d20d127f70e92240d59fe59edc77d59b7031c0e6)

Reply:
I can see where @wence- is coming from, but it's not uncommon to use `dict`s for this purpose, especially within the context of Pandas.

For example, passing a `dict` to the `DataFrame` constructor, or passing a `dict` as the `dtypes` argument to `read_csv`.

Unless this is fundamentally different for some reason I am not seeing?

Reply:
> For example, passing a `dict` to the `DataFrame` constructor, or passing a `dict` as the `dtypes` argument to `read_csv`.

These match up columns to values by the name in the dict. In the former, you happen to get your dataframe columns coming out in the order you wrote the dict but that's fine because you normally reference columns by name. In the latter the order doesn't matter because again the dtypes are tied to the csv columns by name:
```
In [68]: from io import StringIO

In [69]: csv_data = "a,b\n1,2"

In [70]: pd.read_csv(StringIO(csv_data), dtype={"a": str, "b": "float64"})
Out[70]: 
   a    b
0  1  2.0

In [71]: pd.read_csv(StringIO(csv_data), dtype={"b": "float64", "a": str})
Out[71]: 
   a    b
0  1  2.0

In [72]: Out[70] == Out[71]
Out[72]: 
      a     b
0  True  True
```

> Unless this is fundamentally different for some reason I am not seeing?

Here, the `(column_name, dtype)` pairs are only tied to the list of input columns by their index in the datastructure.

To my mind, APIs that accept mappings shouldn't generally require anything about the traversal order, as a principle of least surprise.

Reply:
OK -- I think that's fair

Reply:
Changed to accept a list of tuples. 👍